thousandeyes/util.go

```diff
@@ -95,3 +99,184 @@ func unpackSIPAuthData(i interface{}) thousandeyes.SIPAuthData {
+	for i := 0; i < v.NumField(); i++ {
+		tag := GetJSONKey(t.Field(i))
+		tfName := CamelCaseToUnderscore(tag)
+		if d.HasChange(tfName) {
+			newVal := reflect.ValueOf(FillValue(d.Get(tfName), v.Field(i).Interface()))
+			v.Field(i).Set(newVal)
+		}
+	}
+	d.Partial(false)
+	return structPtr
+}
+
+// ResourceSchemaBuild creates a map of schemas based on the fields
+// of the provided type.
+func ResourceSchemaBuild(referenceStruct interface{}) map[string]*schema.Schema {
+	newSchema := map[string]*schema.Schema{}
+	v := reflect.ValueOf(referenceStruct)
+	t := reflect.TypeOf(referenceStruct)
+
+	for i := 0; i < v.NumField(); i++ {
+		tag := GetJSONKey(t.Field(i))
+		tfName := CamelCaseToUnderscore(tag)
+		if val, ok := schemas[tfName]; ok {
+			newSchema[tfName] = val
+		}
+	}
+	return newSchema
+}
+
+// FillValue translats a value from the Terraform provider framework and
+// translates it to the correct type, based on the type of the target parameter.
+func FillValue(source interface{}, target interface{}) interface{} {
+	// We determine how to interpret the supplied value based on
+	// the type of the target argument.
+	vt := reflect.ValueOf(target)
+	switch vt.Kind() {
+	case reflect.Slice:
+		// When the target is a slice, we create a new slice of the same type,
+		// then recurse with the value of each element.
+		vs := reflect.ValueOf(source)
+		tt := reflect.TypeOf(target)
+		tte := reflect.TypeOf(target).Elem() // The type of items in the slice
+		ntte := reflect.New(tte).Elem()
+		newSlice := reflect.New(tt).Elem()
+		for i := 0; i < vs.Len(); i++ {
+			toAppend := FillValue(vs.Index(i).Interface(), ntte.Interface())
+			appendVal := reflect.ValueOf(toAppend)
+			newSlice = reflect.Append(newSlice, appendVal)
+		}
+		return newSlice.Interface()
+	case reflect.Struct:
+		// When the target is a struct, we assume that the source is a map
+		// containing corresponding values for the struct's fields, then
+		// recurse on each value looked up to get the value to be set.
+		t := reflect.TypeOf(vt.Interface())
+		newStruct := reflect.New(t).Interface()
+		setStruct := reflect.ValueOf(newStruct).Elem()
+		m := source.(map[string]interface{})
+		for i := 0; i < vt.NumField(); i++ {
+			tag := GetJSONKey(t.Field(i))
+			tfName := CamelCaseToUnderscore(tag)
+			if mv, ok := m[tfName]; ok {
+				newVal := FillValue(mv, vt.Field(i).Interface())
+				setStruct.Field(i).Set(reflect.ValueOf(newVal))
+			} else {
+			}
+		}
+		return setStruct.Interface()
+	case reflect.Int:
+		// Values destined to be ints may come to us as strings.
+		if reflect.TypeOf(source).Kind() == reflect.String {
+			i, _ := strconv.Atoi(source.(string))
+			return i
+		}
+
+		return source
+	default:
+		// If we haven't matched one of the above cases, then there
+		// is likely no reason to translate.
+		return source
+	}
+
+	// Or if the above is too tricky...
+	//switch t := target.(type) {
+	//case []thousandeyes.Agents:
+	//return expandAgents(value)
```

Comment:
How hard would this get to maintain? It seems like it could be more explicit and slightly clearer?

Reply:
I agree that this is more straightforward, but my thought is that it actually requires more active maintenance.  Outside of the aspect of repetitive code potentially hiding errors, each type would require changing to keep up with API changes/fixes.  

It probably wouldn't be too bespoke, as anything should just be a permutation of the existing methods for a list or a map.  But it would require active tracking of them to be sure they exist and work.  Even now, we're missing several.  

The benefit of inferring them with the FillValue function is that we get them and any future updates to them without the upkeep.